extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,389 @@
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO: Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    // See https://github.com/ampproject/amphtml/issues/1226
+    onDocumentReady(this.win_, () => {
+      this.documentReadyResolve_();
+    });
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.whenVisible().then(() => {
+      // we block on both viewer visible and when dom ready.
+      return this.documentReadyPromise_;
+    }).then(() => {
+      Object.keys(this.deferRegistry_).forEach((id) => {
+        assert(this.win_.document.getElementById(id) != null,
+            `Did not find amp-user-notification element ${id}`);
+      });
+    });
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @param {string} id
+   * @return {!Promise<string>}
+   */
+  get(id) {
+    assert(!!this.win_.document.getElementById(id),
+        `No Element found with id ${id} ` +
+        'for `userNotification` service request.');
+    return this.getElementDeferById_(id).promise;
+  }
+
+  /**
+   * Register an instance of `amp-user-notification`.
+   * @param {string} id
+   * @param {!UserNotification} userNotification
+   * @package
+   */
+  registerUserNotification(id, userNotification) {
+    const deferred = this.getElementDeferById_(id);
+
+    // Compose the registered notifications into a promise queue
+    // that blocks until one notification is dismissed.
+    this.nextInQueue_ = this.nextInQueue_
+        .then(() => {
+          return userNotification.shouldShow().then((shouldShow) => {
+            if (shouldShow) {
+              return userNotification.show();
+            }
+          });
+        })
+        .then(deferred.resolve)
+        .catch(this.onRejection_.bind(this, id));
+  }
+
+  onRejection_(id, err) {
+    log.error('Notification service failed amp-user-notification', id, err);
+  }
+
+  /**
+   * Retrieve UserNotificationDeferDef object.
+   * @param {string} id
+   * @return {!UserNotificationDeferDef}
+   * @private
+   */
+  getElementDeferById_(id) {
+    return this.createOrReturnDefer_(id);
+  }
+
+  /**
+   * Create an defer if it doesnt exist, else just return the one in the
+   * registry.
+   * @return {!UserNotificationDeferDef}
+   */
+  createOrReturnDefer_(id) {
+    if (this.deferRegistry_[id]) {
+      return this.deferRegistry_[id];
+    }
+
+    let resolve_;
```

Comment:
done. i need to have a different name for the resolve inside the Promise((resolve) => {}) call.